### ISSUE TITLE
Schedule does not seems to be respected

### ISSUE THREAD

**Issue by erwan-joly** [dependabot/dependabot-core#3059]
Hi, I'm trying to use the schedule parameter with time: "18:00" and  timezone: "Pacific/Auckland" but it seems like dependabot doesn't respect the 18:00 for some reasons ? Am I doing something wrong ?
  ```
updates:
  - directory: "/"
    package-ecosystem: nuget
    schedule:
      interval: daily
      time: "18:00"
      timezone: "Pacific/Auckland"
    open-pull-requests-limit: 100
```

**Comment by jurre** [comment]
The schedule is ~best effort, and it's roughly when we _start_ update jobs, it might take a while for PRs to actually be opened. There are also a few other reasons why a PR could be opened, for example a security advisory was published, and when a job has previously failed we also re-run it when a change to the manifest file has been pushed.

So what you're seeing could be explained by one of the above, could you share a bit more details about what you're seeing, and maybe share the repo name so I can check some logs etc?

**Comment by jurre** [comment]
> Those are definitely not security advisory. 

👍 

> The logs in the dependabot change didn't start too far from 6PM (7PM) yesterday 2021-02-02T06:59:06Z

Without knowing the repo it's hard to check up on this, an hour later than scheduled is not great, but it _can_ happen due to a lot of back pressure. Feel free to open a ticket via support so it's easier to share details if you'd like us to dig deeper!

> Additionally All the PR got rebased at about 3PM

By default, PRs get rebased when dependabot detects conflicts, but this can be disabled: https://docs.github.com/en/github/administering-a-repository/configuration-options-for-dependency-updates#rebase-strategy

**Comment by MrChocolatine** [comment]
Since you recommend to search for similar issues before posting one, with my company we have the same issue for one private repository.

We have something like ~30 repositories for Javascript projects, all of them sharing the same configuration (see below) which basically schedules 2 upgrades maximum every week.

But only for one reposityry Dependabot keeps creating upgrades as soon as there is one available. The initial limit has already been processed long time ago. No matter what we do (recreating the file for instance) Dependabot creates an upgrade, at any time.

<details>
<summary>The shared config file</summary>

```yaml
version: 2

updates:

  # Required options
  - package-ecosystem: npm
    directory: "/"
    schedule:
      interval: weekly
      day: saturday
      time: "00:00"
      timezone: Europe/Paris

    # Behaviour of pull requests
    open-pull-requests-limit: 2
    pull-request-branch-name:
      separator: "-"
    commit-message:
      prefix: build
      include: scope

    # Metadata of pull requests
    reviewers:
      - "peopledoc/tribe-js"

    # Control which dependencies are updated
    ignore:
      - dependency-name: "ember-cli"
```

</details>

**Comment by MrChocolatine** [comment]
Since you recommend to search for similar issues before posting one, with my company we have the same issue for one private repository.

We have something like ~30 repositories for Javascript projects, all of them sharing the same configuration (see below) which basically schedules 2 upgrades maximum every week.

But only for one reposityry Dependabot keeps creating upgrades as soon as there is one available. The initial limit has already been processed long time ago. No matter what we do (recreating the file for instance) Dependabot creates an upgrade, at any time.

<details>
<summary>The shared config file</summary>

```yaml
version: 2

updates:

  # Required options
  - package-ecosystem: npm
    directory: "/"
    schedule:
      interval: weekly
      day: saturday
      time: "00:00"
      timezone: Europe/Paris

    # Behaviour of pull requests
    open-pull-requests-limit: 2
    pull-request-branch-name:
      separator: "-"
    commit-message:
      prefix: build
      include: scope

    # Metadata of pull requests
    reviewers:
      - "xxx/yyy"

    # Control which dependencies are updated
    ignore:
      - dependency-name: "ember-cli"
```

</details>

**Comment by mueller-ma** [comment]
I have a PR that I can share here: https://github.com/openhab/openhab-android/pull/2590
Opened at 21:38 MEST, but I configured 04:00:

````yaml
version: 2
updates:
- package-ecosystem: github-actions
  directory: "/"
  schedule:
    interval: daily
    time: "04:00"
    timezone: Europe/Berlin
  open-pull-requests-limit: 99
- package-ecosystem: gradle
  directory: "/"
  schedule:
    interval: daily
    time: "04:00"
    timezone: Europe/Berlin
  open-pull-requests-limit: 99

````

Since there was another PR one minute earlier and none of them have something about security in their release notes, I don't think that's a security relase: https://github.com/openhab/openhab-android/pull/2588

**Comment by guatedude2** [comment]
I'm having the same issue. I setup two repos with the following dependabot.yml configuration:

```yaml
version: 2
updates:
  - package-ecosystem: "npm"
    directory: "/"
    schedule:
      interval: "weekly"
      day: "monday"
    ignore:
      - dependency-name: "*"
        update-types: ["version-update:semver-patch"]
```

For some reason, the PRs get opened right after I close any of the 5 open PRs. The schedule isn't respected at all and PRs are just opened whenever one is closed. Any ideas why that may be?

**Comment by mueller-ma** [comment]
Maybe because of the default pr limit? Can you try to increase it as in my config above?

**Comment by guatedude2** [comment]
Ok I applied the limit, i'll report back with status

**Comment by guatedude2** [comment]
Looks like that just kept on popping PRs up, still not respecting the schedule (weekly on Monday's) today is Sunday

**Comment by guatedude2** [comment]
I had to close them manually:
<img width="746" alt="Screen Shot 2021-05-30 at 2 54 30 PM" src="https://user-images.githubusercontent.com/918429/120116618-f2377100-c156-11eb-90d2-8cd1c73cc9b7.png">

**Comment by samsmithnz** [comment]
I have an example too. This should run daily at 6am, but for some reason - hasn't run for the last 2 days - I know there is a package to update to. Repo: https://github.com/samsmithnz/AzurePipelinesToGitHubActionsConverterWeb/blob/main/.github/dependabot.yml
```
version: 2
updates:
- package-ecosystem: nuget
  directory: "/PipelinesToActions/PipelinesToActions"
  schedule:
    interval: daily
    time: "06:00"
    timezone: America/New_York
  open-pull-requests-limit: 10
  assignees:
  - "samsmithnz"
```
![image](https://user-images.githubusercontent.com/8389039/120922273-2d292f80-c696-11eb-8316-2feceef3f021.png)

**Comment by thabemmz** [comment]
Having the exact same issue as @guatedude2.

## Issue
As soon as we merge 1 PR, Dependabot creates a new one, effectively keeping us busy reviewing Dependabot updates all week.

## What I'd expect
I'd expect Dependabot to respect the schedule and only offer updates on Sunday, 19:00 with a maximum of 10 Dependabot PR's open at the same time

As a work-around, we have disabled patch and minor updates for now, but this is of course FAR from ideal.

We have a private repository.

Our `dependabot.yml`:
```
version: 2
updates:
  - package-ecosystem: npm
    directory: "/"
    schedule:
      day: sunday
      interval: weekly
      time: "19:00"
      timezone: Europe/Amsterdam
    open-pull-requests-limit: 10
```

**Comment by Narnach** [comment]
The behavior as described by others still persists.

It's almost as if Dependabot has its own internal queue of dependency updates (which is filled on the scheduled interval), but as soon as we merge or close a Dependabot PR, it will create the next one from its queue.

The desired behavior is that it only creates new (non-security update) PRs at the scheduled time, and not creates more PRs out-of-bound after closing previous PRs. The one exception to this would be when Dependabot is newly added to a project, or a new package ecosystem is added to the Dependabot config file.

The reason we use Dependabot is to allocate a roughly fixed amount of time each week to perform updates, to have an upper bound on the amount of time we spend on our maintenance and technical debt each week. Right now the out-of-bound Dependabot PRs clutter our PR list, which adds an annoyance factor to an otherwise awesome product.

**Comment by jurre** [comment]
What I _think_ is happening here, is that dependabot is triggered when your manifest files (package.json for example) change, and kicks off a new job. Normally when this happens during the week for a manual change to those files, dependabot will not do anything because the open PR limit is reached, however when this happens when a PR is merged, once that job is finished, Dependabot will notice a "free" slot in that limit, and open a PR.

I get this is annoying and unexpected, I'm just not immediately sure of a good way to fix it. We might be able to disable the behavior of re-running dependabot on a manifest file change when we notice that the person who changed it was Dependabot, but that might be slightly tricky, as a merged PR will show the author as the person that merged 🤔

**Comment by Narnach** [comment]
Thank @jurre, that looks in line with what @asciimike describes in https://github.com/dependabot/dependabot-core/issues/3975#issuecomment-868847404

We define the scheduled interval specifically to have one weekly moment where new Dependabot PRs are created. It's on Sunday night for us, so CI can run as long as it needs and we can review dependency PRs on Monday morning. We cherry pick the Dependabot PRs into one new branch/PR and perform some additional work (applying new linting rules, grabbing all `@types/*` updates if one specific type was suggested by Dependabot, etc). This is mostly due to having strict PR review procedures, so only having one PR review for all dependencies is the most efficient for us.

We thought that we could control the behavior via `open-pull-requests-limit`, combined with `schedule`: limit how often Dependabot checks dependencies, and limit how many PRs it creates when it checks.

What you describe re: creating new Dependabot PRs for existing package ecosystems in response to manifest file changes is _explicitly_ an anti-goal for us. Looking at manifest changes to close or update existing PRs _is_ expected behavior. Security updates already follow their own out-of-bounds schedule, so that's no reason for the behavior either.

What's the reasoning for looking at manifest file changes and opening new PRs, in context of there being a schedule _and_ PR limits? If your intention behind the PR limits to have an unlimited "queue" of PRs Dependabot wants to create, but to simply limit the number of PRs which can exist at the same time? In this case maybe we need an additional `max-internal-pr-queue-size: <integer>` or a `can-create-unscheduled-prs: <boolean>`, or `recheck-on-manifest-change: <boolean>` setting to control this behavior so teams can configure this behavior in a way they prefer.

In the PR I mentioned I saw that the team raising the issue switched to Renovate over this. I can tell you that members of our team have mentioned them as an alternative to look at over this exact issue as well. Personally I like to see what we can do to help you solve the issue and stay on Dependabot, so I'm happy to help to try to think of solutions.

**Comment by jurre** [comment]
@Narnach could you share the repo this is happening on? I'd like to run some checks

**Comment by Narnach** [comment]
@jurre it's for reisbalans/reisbalans, a private repo

**Comment by jurre** [comment]
Thanks!

After digging deeper I think an update is only supposed to happen if the last update failed. The reasons there being that the failure is possibly resolved after a manifest file change.

So this shouldn't happen _every_ time you merge a PR 🤔 

I do see a few update jobs that have errors in the logs for that repository, could that explain what's happened? Or does it _consistently_ happen every time even when there are no errors?

If you can share the pull request number for one of the PRs that was created after one of these merges that'd help a lot (so, `github.com/org/repo/pulls/<number>`

**Comment by Narnach** [comment]
Dependabot consistently fails the final step of the linting check, because it tries to report annotated feedback to Github. I think Dependabot does not have sufficient rights to annotate our code, hence that job fails (`Error trying to create GitHub check for Prettier: Received status code 403`).

It works when our team members initiate the check. We figured it was due to Github tightening security measures over the last few months. Because we create a new PR for the actual merge, it's been a low priority to look into it.

The most recent dependency PR we merged was:

* https://github.com/reisbalans/reisbalans/pull/6041

It created these PRs in response:

* https://github.com/reisbalans/reisbalans/pull/6045
* https://github.com/reisbalans/reisbalans/pull/6046
* https://github.com/reisbalans/reisbalans/pull/6047
* https://github.com/reisbalans/reisbalans/pull/6048
* https://github.com/reisbalans/reisbalans/pull/6049
* https://github.com/reisbalans/reisbalans/pull/6050

**Comment by jurre** [comment]
> Dependabot consistently fails the final step of the linting check, because it tries to report annotated feedback to Github. I think Dependabot does not have sufficient rights to annotate our code, hence that job fails (Error trying to create GitHub check for Prettier: Received status code 403).

This is a CI check you're referring to? That should be fine and ignored, we only care about the dependabot itself failing (you should be able to see this on https://github.com/reisbalans/reisbalans/network/updates).

> The most recent dependency PR we merged was:
>
> https://github.com/reisbalans/reisbalans/pull/6041

I don't have access to this repo, but in the Dependabot DB I cannot find a reference to that PR, so am I right in assuming this was not created by us? Was the dependabot config file changed in that PR by any chance?

**Comment by Narnach** [comment]
Sorry, I thought you meant CI check logs 😅

The network updates check failures appear to be for hitting the limit on open PRs: package.json has a limit of 5 that's reached, and we've got a secondary Gemfile in a subdirectory for an embedded project which has been set to limit 0 to ignore it (seemed the best way to get Dependabot to ignore a manifest file 🤷 )

PR 6041 was indeed created by me, and is the aggregate dependency update PR for this week. I _did_ update the Dependabot config there to make Dependabot ignore a gem which is pinned on a specific version (and will be phased out rather than updated soon), but the `schedule` segment did not get changed, so from our perspective a batch of new PRs was not intended or expected 😉

**Comment by jurre** [comment]
> The network updates check failures appear to be for hitting the limit on open PRs: package.json has a limit of 5 that's reached, and we've got a secondary Gemfile in a subdirectory for an embedded project which has been set to limit 0 to ignore it

That should be fine and not count as an "error" state, that's just Dependabot working as intended and not opening more PRs than what's configured.

>  I did update the Dependabot config there to make Dependabot ignore a gem which is pinned on a specific version (and will be phased out rather than updated soon), but the schedule segment did not get changed, so from our perspective a batch of new PRs was not intended or expected

Ah, _any_ change to the config file triggers updates immediately. Our assumption is that users will want these changes applied immediately in most cases.

I think that explains what may have happened here, or can you think of other cases where this definitely was not the culprit?

The only thing I can recommend right now is to separate changes to the config file out and apply them before merging/closing any of Dependabot's PRs.

This behavior is documented [here](https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/about-dependabot-version-updates#frequency-of-dependabot-pull-requests) fwiw, and I think it makes sense for most users. I suppose we could add an option to disable this, but to be honest I don't think our small team currently has the bandwidth to tackle it any time soon.

**Comment by Narnach** [comment]
> Ah, _any_ change to the config file triggers updates immediately. Our assumption is that users will want these changes applied immediately in most cases.
> 
> I think that explains what may have happened here, or can you think of other cases where this definitely was not the culprit?

Checking back I see evidence to support this theory (internal PRs):

* https://github.com/reisbalans/reisbalans/pull/5967, a regular dependency PR, did not change dependabot.yml and appears to _not_ have created new PRs
* https://github.com/reisbalans/reisbalans/pull/5762 changed dependabot.yml to change the versioning strategy, but that was intended to apply on the next scheduled check, so the new PRs were unexpected for us

> The only thing I can recommend right now is to separate changes to the config file out and apply them before merging/closing any of Dependabot's PRs.
>
> This behavior is documented [here](https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/about-dependabot-version-updates#frequency-of-dependabot-pull-requests) fwiw, and I think it makes sense for most users. I suppose we could add an option to disable this, but to be honest I don't think our small team currently has the bandwidth to tackle it any time soon.

Having an option to disable unscheduled non-security PR creation would probably be the most easily understandable option for users.

I did a quick search in the code, but scheduling does not appear to be part of the open source code, so I don't think I can help out with a PR for this.

I hope your team manages to pick it up at some point. I'll see if our increased understanding of Dependabot's behavior makes it less annoying for us.

Thanks for the informative conversation!

**Comment by jurre** [comment]
Thanks @Narnach 

> Having an option to disable unscheduled non-security PR creation would probably be the most easily understandable option for users.

cc @asciimike, for future reference

> I did a quick search in the code, but scheduling does not appear to be part of the open source code, so I don't think I can help out with a PR for this.

Yeah that's right, this is all part of the services we use to run Dependabot at GitHub which are all private.

> I hope your team manages to pick it up at some point. I'll see if our increased understanding of Dependabot's behavior makes it less annoying for us.

I hope it does! 💛 

> Thanks for the informative conversation!

Thanks for helping me figure out what happened here!

**Comment by Narnach** [comment]
@jurre I've just processed this week's dependency updates, but after merging my PR we got 5 new Dependabot PRs. My PR did _not_ contain changes to `.github/dependabot.yml`. This seems to (at least partially) refute the hypothesis that changing the config file was the _only_ cause of new PRs outside of scheduled times.

My dependencies PR:

* https://github.com/reisbalans/reisbalans/pull/6066

After merging the PR, we got these additional PRs from Dependabot:

* https://github.com/reisbalans/reisbalans/pull/6068
* https://github.com/reisbalans/reisbalans/pull/6069
* https://github.com/reisbalans/reisbalans/pull/6070
* https://github.com/reisbalans/reisbalans/pull/6071
* https://github.com/reisbalans/reisbalans/pull/6072

We _did_ get a PR for an update to a Github Action, so that resulted in some of the workflow files in `.github/workflows/*.yml` getting updated. Thinking out loud: could it be that you have code somewhere checking for changes to `.github/**/*.yml` instead of only `.github/dependabot.yml`? 🤔

**Comment by jurre** [comment]
> Thinking out loud: could it be that you have code somewhere checking for changes to .github/**/*.yml instead of only .github/dependabot.yml?

I don't think so, the code in question checks against this constant:

```ruby
CONFIG_FILE_PATHS = %w(.github/dependabot.yml .github/dependabot.yaml).freeze
```

But thanks for these links, I'll try to find some time to figure out what happened here!

**Comment by rafalyesware** [comment]
One side-effect of this I've seen a number of times over the last 2-3 days, is that when dependabot creates a new PR to replace a closed one, sometimes those PRs appear to use stale state.  Here's an example (unfortunately in a private repo, but at least illustrative):

Thu Jul 29 19:25:59 2021 +0000, I merge a dependabot PR to "Bump ngx-infinite-scroll from 9.1.0 to 10.0.1".
Thu Jul 29 19:29:55 2021 +0000, dependabot opens a PR to "Bump jasmine-core from 3.7.1 to 3.8.0".  Inspecting the diffs (and the git log), this PR is based on the state of `master` **previous** to the PR above, and also includes the upgrade to "ngx-infinite-scroll" again.

While this is probably somewhat harmless in many cases, there are a couple of things that I find really annoying about this:

* Everytime I've looked at one of these PRs that are based off stale state, I have to double-check myself and the state of the repo, wondering why this is also upgrading a package I already had upgraded previously, so those upgrades take up a lot more mental bandwidth.
* Secondly, it breaks the rule of one-package-upgrade-per-PR, which makes diagnosing what broke our production app that much more difficult.  Maybe this can only happen in the case when an upgrade to another package has just been merged, but it makes me lose faith in dependabot when the PR says "Bump X", but the contents are actually "Bump X & Y".

**Comment by asciimike** [comment]
I've got an ongoing conversation with the PR team about the "Dependabot PRs are in a stale state", though unfortunately it's not moving quickly. It's not specific to Dependabot, but Dependabot users happen to hit it a lot more frequently than the average user. Your comment about rebasing to resolve the "bump x and y" is the right course of action, though I agree it's a pain :/

**Comment by jsyrjala** [comment]
I am seeing this same behavior with gradle projects. Schedule is not respected, and often new PRs get created soon after the old ones are merged.

**Comment by rubencodes** [comment]
Same issue here for `npm` updates—new PRs get made right after the last one is closed, completely defeating the purpose of the "schedule" option. My team was really hoping to use dependabot for keeping things up to date at a steady pace, but this is a non-starter 😞

**Comment by martinvium** [comment]
Same problem, every time I close a PR a new one get's created.

Config looks something like this:

```
- package-ecosystem: bundler
  directory: "/"
  schedule:
    interval: weekly
    day: wednesday
    time: "06:00"
    timezone: Europe/Copenhagen
  open-pull-requests-limit: 5
```

**Comment by MrChocolatine** [comment]
Incredible how maintainers keep giving us feedback on this incorrect behaviour.

**Comment by ayg-qliktech** [comment]
Hi. Same problem. Even though we have specified "weekly" interval. It happens daily instead. Schedule interval deeming itself completely useless in this case. Snippet from our config:

- package-ecosystem: npm
  directory: "/"
  schedule:
    interval: weekly
    day: "friday"
    time: "12:30"
    timezone: US/Eastern

**Comment by MrChocolatine** [comment]
Folks, just use Renovate instead.
https://github.com/renovatebot/renovate

I feel like Dependabot is becoming an abandonware.

**Comment by janhartigan** [comment]
@MrChocolatine thanks for sharing that, I've never seen that before so I'll evaluate it for our team.

Out of curiosity, why do you say Dependabot is becoming abandonware? Obviously this issue is concerning, but wondering if you have some more info.

**Comment by MrChocolatine** [comment]
I may be completely wrong, I am just a simple user.

> why do you say Dependabot is becoming abandonware

**I am not saying it is becoming an abandonware**, but I feel like.
Given the lack of news, how the product does not seem to improve over the years...

Again, it's a personal feeling, perhaps I am expecting too much for this service.

**Comment by MrChocolatine** [comment]
I am glad to be wrong then, thank you for this message.

**Comment by raphael-theriault-swi** [comment]
Would it be possible to at least get a configuration option to ignore changes to manifest files and only trigger either on schedule or when the dependabot config is changed ? As it stands the schedule feature seems like it's pretty much useless

**Comment by abdulapopoola** [comment]
@raphael-theriault-swi ; @deivid-rodriguez is working on this and we hope to have this resolved this week :). Thanks for being patient and raising this up.

**Comment by deivid-rodriguez** [comment]
Hei! We deployed an internal fix for this. Please let us know if you receive any more PRs out of schedule from now on. (Note that when you edit your `.dependabot.yml` file, that will still retrigger a dependabot run, and that has not changed).

**Comment by kazazor** [comment]
@deivid-rodriguez I;'m still getting weird schedule problems with dependabot:
```yml
version: 2
updates:
  - package-ecosystem: "npm"
    directory: "/"
    schedule:
      interval: "monthly"

```

While configured to monthly, it runs several times a month.

Any input on that?

**Comment by deivid-rodriguez** [comment]
The only situation when Dependabot runs out of schedule now is when the `.dependabot.yml` changes are pushed. If you're not experiencing this, I suggest you open a [support ticket](https://support.github.com) so that your particular situation can be investigated.

**Comment by LesnyRumcajs** [comment]
@deivid-rodriguez 

The schedule is still not respected for `cargo`. Should this issue be closed? This happens across a few repositories, configured with a weekly schedule.

[chainsafe/forest](https://github.com/ChainSafe/forest) with this [weekly schedule](https://github.com/ChainSafe/forest/blob/0f4578eb2292e1f0f69d2d5c3e9e8328132b0cd4/.github/dependabot.yml) fires significantly more often than it should.

![image](https://github.com/dependabot/dependabot-core/assets/9642092/88c9e02e-6c8c-4408-9b4d-77950b7494b1)

It's exactly the same thing in [chainsafe/fil-actor-states](https://github.com/ChainSafe/fil-actor-states).

I also sent this to GH support, but perhaps other people have the same problem. If so, I'm not convinced closing the issue is the right way to go.
